File: integration-tests-java/src/main/java/io/micronaut/jimmer/java/it/repository/TreeNodeRepository.java
package io.micronaut.jimmer.java.it.repository;

import io.micronaut.jimmer.java.it.entity.Tables;
import io.micronaut.jimmer.java.it.entity.TreeNode;
import io.micronaut.jimmer.java.it.entity.dto.TreeNodeDetailView;
import io.micronaut.jimmer.repository.JRepository;
import io.micronaut.jimmer.repository.annotation.Repository;
import java.util.List;

@Repository
public interface TreeNodeRepository extends JRepository<TreeNode, Long> {

    default List<TreeNodeDetailView> infiniteRecursion(Long parentId) {
        return sql().createQuery(Tables.TREE_NODE_TABLE)
                .where(Tables.TREE_NODE_TABLE.parentId().eq(parentId))
                .select(Tables.TREE_NODE_TABLE.fetch(TreeNodeDetailView.class))
                .execute();
    }
}

File: integration-tests-kotlin/src/main/kotlin/io/micronaut/jimmer/kotlin/it/repository/TreeNodeRepository.kt
package io.micronaut.jimmer.kotlin.it.repository

import io.micronaut.jimmer.kotlin.it.entity.TreeNode
import io.micronaut.jimmer.kotlin.it.entity.dto.TreeNodeDetailView
import io.micronaut.jimmer.kotlin.it.entity.parentId
import io.micronaut.jimmer.repository.KRepository
import io.micronaut.jimmer.repository.annotation.Repository
import org.babyfish.jimmer.sql.kt.ast.expression.`eq?`

@Repository
interface TreeNodeRepository : KRepository<TreeNode, Long> {
    fun infiniteRecursion(parentId: Long?): List<TreeNodeDetailView> =
        sql
            .createQuery(TreeNode::class) {
                where(table.parentId `eq?` parentId)
                select(table.fetch(TreeNodeDetailView::class))
            }.execute()
}
